zerostash/src/commands/mount.rs

```diff
@@ -0,0 +1,279 @@
+//! `mount` subcommand
+
+use std::ffi::OsStr;
+use std::path::PathBuf;
+use std::sync::{mpsc, Arc, Mutex};
+use std::time::{Duration, SystemTime};
+use std::{collections::HashMap, path::Path};
+
+use std::io::Result;
+
+use crate::prelude::*;
+use fuse_mt::*;
+use infinitree::object::Reader;
+use infinitree::Infinitree;
+use nix::libc;
+use zerostash_files::{restore, Entry, Files};
+
+#[derive(Command, Debug)]
+pub struct Mount {
+    #[clap(flatten)]
+    stash: StashArgs,
+
+    #[clap(flatten)]
+    options: restore::Options,
+
+    /// The location the filesytem will be mounted on
+    #[clap(short = 'T', long = "target")]
+    mount_point: String,
+}
+
+#[cfg(unix)]
+#[async_trait]
+impl AsyncRunnable for Mount {
+    /// Start the application.
+    async fn run(&self) {
+        let stash = self.stash.open();
+
+        if let Err(e) = mount(stash, &self.options, &self.mount_point) {
+            panic!("Error = {}", e)
+        }
+    }
+}
+
+fn mount(
+    stash: Infinitree<Files>,
+    options: &restore::Options,
+    mountpoint: &str,
+) -> anyhow::Result<()> {
+    let (tx, finished) = mpsc::sync_channel(2);
+    let destroy_tx = tx.clone();
+    ctrlc::set_handler(move || tx.send(()).expect("Could not send signal on channel."))
+        .expect("Error setting Ctrl-C handler");
+
+    let filesystem = SimpleFs::open(stash, options, destroy_tx).unwrap();
+    let fuse_args = [OsStr::new("-o"), OsStr::new("fsname=zerostash")];
+
+    let fs = fuse_mt::FuseMT::new(filesystem, 1);
+
+    // Mount the filesystem.
+    let handle = spawn_mount(fs, mountpoint, &fuse_args[..])?;
+
+    // Wait until we are done.
+    finished.recv().expect("Could not receive from channel.");
+
+    // Ensure the filesystem is unmounted.
+    handle.join();
+
+    Ok(())
+}
+
+type DirPath = PathBuf;
+type FilePath = PathBuf;
+type FileContent = Vec<u8>;
+type Metadata = Arc<Entry>;
+
+struct SimpleFs {
+    destroy_tx: mpsc::SyncSender<()>,
+    stash: Infinitree<Files>,
+    file_map: HashMap<FilePath, Metadata>,
+    file_content: Mutex<HashMap<FilePath, FileContent>>,
+    dir_map: HashMap<DirPath, Vec<DirectoryEntry>>,
+    file_attr_memory: Mutex<HashMap<PathBuf, FileAttr>>,
+}
+
+fn add_dir_to_map(
+    dir_map: &mut HashMap<DirPath, Vec<DirectoryEntry>>,
+    path: &Path,
+    kind: FileType,
+) {
+    let name = path
+        .file_name()
+        .expect("All files have filenames")
+        .to_owned();
+
+    let parent = path
+        .parent()
+        .expect("Paths should have parents")
+        .to_path_buf();
+
+    if !dir_map.contains_key(&parent) {
+        add_dir_to_map(dir_map, &parent, FileType::Directory);
+    }
+
+    dir_map
+        .entry(parent)
+        .or_default()
+        .push(DirectoryEntry { name, kind });
+}
```

Comment:
This looks like a good start, but may be interesting to think about creating a directory-directory in the `Files` struct that records relevant information in a separate `HashMap`. This means you could crawl/iterate/cache those `infinitree` fields directly on the index instead of having to transform them into memory.

Reply:
Wouldn't I still need to transform it into memory because I have to return a `Vec<DirectoryEntry>` in `readdir`.

Reply:
I think the difference is that in that case it would be trivial to construct these on the fly due to having a cache already in place. It seems to me that now the difficulty is that you would need to crawl the entire stash to find anything belonging to the given directory, which _is_ expensive.

Building the `DirectoryEntry` instance for the specific requested directory should be fairly straightforward, as you would just be able to transform the readily available cache entry, ie, making it `O(n)` over directory size, as opposed to `O(n)` over stash size.